// puzzlescript/lib/parser/astTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LEVEL_TYPE = exports.WIN_CONDITION_TYPE = exports.COMMAND_TYPE = exports.BRACKET_TYPE = exports.RULE_TYPE = exports.SOUND_TYPE = exports.TILE_TYPE = exports.SPRITE_TYPE = exports.COLOR_TYPE = exports.SOUND_SPRITE_EVENT = exports.SOUND_SPRITE_DIRECTION = exports.SOUND_WHEN = exports.TILE_MODIFIER = exports.RULE_MODIFIER = void 0;
var RULE_MODIFIER;
(function (RULE_MODIFIER) {
    RULE_MODIFIER["RANDOM"] = "RANDOM";
    RULE_MODIFIER["UP"] = "UP";
    RULE_MODIFIER["DOWN"] = "DOWN";
    RULE_MODIFIER["LEFT"] = "LEFT";
    RULE_MODIFIER["RIGHT"] = "RIGHT";
    RULE_MODIFIER["VERTICAL"] = "VERTICAL";
    RULE_MODIFIER["HORIZONTAL"] = "HORIZONTAL";
    RULE_MODIFIER["ORTHOGONAL"] = "ORTHOGONAL";
    RULE_MODIFIER["PERPENDICULAR"] = "PERPENDICULAR";
    RULE_MODIFIER["PARALLEL"] = "PARALLEL";
    RULE_MODIFIER["MOVING"] = "MOVING";
    RULE_MODIFIER["LATE"] = "LATE";
    RULE_MODIFIER["RIGID"] = "RIGID";
})(RULE_MODIFIER = exports.RULE_MODIFIER || (exports.RULE_MODIFIER = {}));
var TILE_MODIFIER;
(function (TILE_MODIFIER) {
    TILE_MODIFIER["NO"] = "NO";
    TILE_MODIFIER["LEFT"] = "LEFT";
    TILE_MODIFIER["RIGHT"] = "RIGHT";
    TILE_MODIFIER["UP"] = "UP";
    TILE_MODIFIER["DOWN"] = "DOWN";
    TILE_MODIFIER["RANDOMDIR"] = "RANDOMDIR";
    TILE_MODIFIER["RANDOM"] = "RANDOM";
    TILE_MODIFIER["STATIONARY"] = "STATIONARY";
    TILE_MODIFIER["MOVING"] = "MOVING";
    TILE_MODIFIER["ACTION"] = "ACTION";
    TILE_MODIFIER["VERTICAL"] = "VERTICAL";
    TILE_MODIFIER["HORIZONTAL"] = "HORIZONTAL";
    TILE_MODIFIER["PERPENDICULAR"] = "PERPENDICULAR";
    TILE_MODIFIER["PARALLEL"] = "PARALLEL";
    TILE_MODIFIER["ORTHOGONAL"] = "ORTHOGONAL";
    TILE_MODIFIER["ARROW_ANY"] = "ARROW_ANY";
})(TILE_MODIFIER = exports.TILE_MODIFIER || (exports.TILE_MODIFIER = {}));
var SOUND_WHEN;
(function (SOUND_WHEN) {
    SOUND_WHEN["RESTART"] = "RESTART";
    SOUND_WHEN["UNDO"] = "UNDO";
    SOUND_WHEN["TITLESCREEN"] = "TITLESCREEN";
    SOUND_WHEN["STARTGAME"] = "STARTGAME";
    SOUND_WHEN["STARTLEVEL"] = "STARTLEVEL";
    SOUND_WHEN["ENDLEVEL"] = "ENDLEVEL";
    SOUND_WHEN["ENDGAME"] = "ENDGAME";
    SOUND_WHEN["SHOWMESSAGE"] = "SHOWMESSAGE";
    SOUND_WHEN["CLOSEMESSAGE"] = "CLOSEMESSAGE";
})(SOUND_WHEN = exports.SOUND_WHEN || (exports.SOUND_WHEN = {}));
var SOUND_SPRITE_DIRECTION;
(function (SOUND_SPRITE_DIRECTION) {
    SOUND_SPRITE_DIRECTION["UP"] = "UP";
    SOUND_SPRITE_DIRECTION["DOWN"] = "DOWN";
    SOUND_SPRITE_DIRECTION["LEFT"] = "LEFT";
    SOUND_SPRITE_DIRECTION["RIGHT"] = "RIGHT";
    SOUND_SPRITE_DIRECTION["HORIZONTAL"] = "HORIZONTAL";
    SOUND_SPRITE_DIRECTION["VERTICAL"] = "VERTICAL";
})(SOUND_SPRITE_DIRECTION = exports.SOUND_SPRITE_DIRECTION || (exports.SOUND_SPRITE_DIRECTION = {}));
var SOUND_SPRITE_EVENT;
(function (SOUND_SPRITE_EVENT) {
    SOUND_SPRITE_EVENT["CREATE"] = "CREATE";
    SOUND_SPRITE_EVENT["DESTROY"] = "DESTROY";
    SOUND_SPRITE_EVENT["CANTMOVE"] = "CANTMOVE";
})(SOUND_SPRITE_EVENT = exports.SOUND_SPRITE_EVENT || (exports.SOUND_SPRITE_EVENT = {}));
var COLOR_TYPE;
(function (COLOR_TYPE) {
    COLOR_TYPE["HEX8"] = "HEX8";
    COLOR_TYPE["HEX6"] = "HEX6";
    COLOR_TYPE["NAME"] = "NAME";
})(COLOR_TYPE = exports.COLOR_TYPE || (exports.COLOR_TYPE = {}));
var SPRITE_TYPE;
(function (SPRITE_TYPE) {
    SPRITE_TYPE["NO_PIXELS"] = "NO_PIXELS";
    SPRITE_TYPE["WITH_PIXELS"] = "WITH_PIXELS";
})(SPRITE_TYPE = exports.SPRITE_TYPE || (exports.SPRITE_TYPE = {}));
var TILE_TYPE;
(function (TILE_TYPE) {
    TILE_TYPE["SIMPLE"] = "LEGEND_ITEM_SIMPLE";
    TILE_TYPE["OR"] = "LEGEND_ITEM_OR";
    TILE_TYPE["AND"] = "LEGEND_ITEM_AND";
})(TILE_TYPE = exports.TILE_TYPE || (exports.TILE_TYPE = {}));
var SOUND_TYPE;
(function (SOUND_TYPE) {
    SOUND_TYPE["WHEN"] = "SOUND_WHEN";
    SOUND_TYPE["SFX"] = "SOUND_SFX";
    SOUND_TYPE["SPRITE_DIRECTION"] = "SOUND_SPRITE_DIRECTION";
    SOUND_TYPE["SPRITE_MOVE"] = "SOUND_SPRITE_MOVE";
    SOUND_TYPE["SPRITE_EVENT"] = "SOUND_SPRITE_EVENT";
})(SOUND_TYPE = exports.SOUND_TYPE || (exports.SOUND_TYPE = {}));
var RULE_TYPE;
(function (RULE_TYPE) {
    RULE_TYPE["GROUP"] = "RULE_GROUP";
    RULE_TYPE["LOOP"] = "RULE_LOOP";
    RULE_TYPE["SIMPLE"] = "RULE_SIMPLE";
})(RULE_TYPE = exports.RULE_TYPE || (exports.RULE_TYPE = {}));
var BRACKET_TYPE;
(function (BRACKET_TYPE) {
    BRACKET_TYPE["SIMPLE"] = "BRACKET_SIMPLE";
    BRACKET_TYPE["ELLIPSIS"] = "BRACKET_ELLIPSIS";
})(BRACKET_TYPE = exports.BRACKET_TYPE || (exports.BRACKET_TYPE = {}));
var COMMAND_TYPE;
(function (COMMAND_TYPE) {
    COMMAND_TYPE["MESSAGE"] = "COMMAND_MESSAGE";
    COMMAND_TYPE["AGAIN"] = "COMMAND_AGAIN";
    COMMAND_TYPE["CANCEL"] = "COMMAND_CANCEL";
    COMMAND_TYPE["CHECKPOINT"] = "COMMAND_CHECKPOINT";
    COMMAND_TYPE["RESTART"] = "COMMAND_RESTART";
    COMMAND_TYPE["WIN"] = "COMMAND_WIN";
    COMMAND_TYPE["SFX"] = "COMMAND_SFX";
})(COMMAND_TYPE = exports.COMMAND_TYPE || (exports.COMMAND_TYPE = {}));
var WIN_CONDITION_TYPE;
(function (WIN_CONDITION_TYPE) {
    WIN_CONDITION_TYPE["SIMPLE"] = "WINCONDITION_SIMPLE";
    WIN_CONDITION_TYPE["ON"] = "WINCONDITION_ON";
})(WIN_CONDITION_TYPE = exports.WIN_CONDITION_TYPE || (exports.WIN_CONDITION_TYPE = {}));
var LEVEL_TYPE;
(function (LEVEL_TYPE) {
    LEVEL_TYPE["MESSAGE"] = "LEVEL_MESSAGE";
    LEVEL_TYPE["MAP"] = "LEVEL_MAP";
})(LEVEL_TYPE = exports.LEVEL_TYPE || (exports.LEVEL_TYPE = {}));
