// puzzlescript/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HexColor = exports.Dimension = exports.TableUI = exports.shouldTick = exports.pollingPromise = exports.MESSAGE_TYPE = exports.A11Y_MESSAGE_TYPE = exports.spritesThatInteractWithPlayer = exports.GameSprite = exports.CollisionLayer = exports._debounce = exports.INPUT_BUTTON = exports.EmptyGameEngineHandler = exports._flatten = exports.saveCoverageFile = exports.LEVEL_TYPE = exports.logger = exports.Serializer = exports.BaseUI = exports.RULE_DIRECTION = exports.GameData = exports.Cell = exports.GameEngine = exports.Parser = void 0;
const engine_1 = require("./engine");
Object.defineProperty(exports, "Cell", { enumerable: true, get: function () { return engine_1.Cell; } });
Object.defineProperty(exports, "GameEngine", { enumerable: true, get: function () { return engine_1.GameEngine; } });
const game_1 = require("./models/game");
Object.defineProperty(exports, "GameData", { enumerable: true, get: function () { return game_1.GameData; } });
const parser_1 = __importDefault(require("./parser/parser"));
exports.Parser = parser_1.default;
const serializer_1 = __importDefault(require("./parser/serializer"));
exports.Serializer = serializer_1.default;
const base_1 = __importDefault(require("./ui/base"));
exports.BaseUI = base_1.default;
const util_1 = require("./util");
Object.defineProperty(exports, "RULE_DIRECTION", { enumerable: true, get: function () { return util_1.RULE_DIRECTION; } });
// Semi-public API (used by the CLI)
const logger_1 = require("./logger");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_1.logger; } });
const astTypes_1 = require("./parser/astTypes");
Object.defineProperty(exports, "LEVEL_TYPE", { enumerable: true, get: function () { return astTypes_1.LEVEL_TYPE; } });
const recordCoverage_1 = require("./recordCoverage");
Object.defineProperty(exports, "saveCoverageFile", { enumerable: true, get: function () { return recordCoverage_1.saveCoverageFile; } });
const util_2 = require("./util");
Object.defineProperty(exports, "_flatten", { enumerable: true, get: function () { return util_2._flatten; } });
Object.defineProperty(exports, "EmptyGameEngineHandler", { enumerable: true, get: function () { return util_2.EmptyGameEngineHandler; } });
Object.defineProperty(exports, "INPUT_BUTTON", { enumerable: true, get: function () { return util_2.INPUT_BUTTON; } });
const collisionLayer_1 = require("./models/collisionLayer");
Object.defineProperty(exports, "CollisionLayer", { enumerable: true, get: function () { return collisionLayer_1.CollisionLayer; } });
const tile_1 = require("./models/tile");
Object.defineProperty(exports, "GameSprite", { enumerable: true, get: function () { return tile_1.GameSprite; } });
const util_3 = require("./util");
Object.defineProperty(exports, "_debounce", { enumerable: true, get: function () { return util_3._debounce; } });
Object.defineProperty(exports, "spritesThatInteractWithPlayer", { enumerable: true, get: function () { return util_3.spritesThatInteractWithPlayer; } });
const rule_1 = require("./models/rule");
Object.defineProperty(exports, "A11Y_MESSAGE_TYPE", { enumerable: true, get: function () { return rule_1.A11Y_MESSAGE_TYPE; } });
const util_4 = require("./util");
Object.defineProperty(exports, "MESSAGE_TYPE", { enumerable: true, get: function () { return util_4.MESSAGE_TYPE; } });
Object.defineProperty(exports, "pollingPromise", { enumerable: true, get: function () { return util_4.pollingPromise; } });
Object.defineProperty(exports, "shouldTick", { enumerable: true, get: function () { return util_4.shouldTick; } });
const table_1 = __importDefault(require("./ui/table"));
exports.TableUI = table_1.default;
const colors_1 = require("./models/colors");
Object.defineProperty(exports, "HexColor", { enumerable: true, get: function () { return colors_1.HexColor; } });
const metadata_1 = require("./models/metadata");
Object.defineProperty(exports, "Dimension", { enumerable: true, get: function () { return metadata_1.Dimension; } });
